examples/bq_file_load_benchmark/benchmark_tools/tables_processor.py

```diff
@@ -0,0 +1,165 @@
+        duplicate_benchmark_tables(bool): Boolean value to determine what to
+            do if a benchmark table already exists for a given file
+            combination. If True, TableProcessor knows to create another
+            benchmark table with the same combination to increase the
+            number of results for accuracy. If not, TablesProcessor knows
+            to only create a benchmark table for a given combination if one
+            has not yet been created.
+        file_params(dict): Dictionary containing each file parameter and
+            its possible values.
+        bq_logs_dataset(str): Name of dataset hold BQ logs table.
+
+    """
+
+    def __init__(
+            self,
+            bq_project,
+            gcs_project,
+            staging_project,
+            staging_dataset_id,
+            dataset_id,
+            bucket_name,
+            results_table_name,
+            results_table_dataset_id,
+            duplicate_benchmark_tables,
+            file_params,
+            bq_logs_dataset,
+    ):
+        self.bq_project = bq_project
+        self.gcs_project = gcs_project
+        self.staging_project = staging_project
+        self.staging_dataset_id = staging_dataset_id
+        self.dataset_id = dataset_id
+        self.bucket_name = bucket_name
+        self.bucket_util = bucket_util.BucketUtil(
+            bucket_name=self.bucket_name,
+            project_id=self.gcs_project,
+            file_params=file_params,
+        )
+        self.results_table_name = results_table_name
+        self.results_table_dataset_id = results_table_dataset_id
+        self.duplicate_benchmark_tables = duplicate_benchmark_tables
+        self.bq_logs_dataset = bq_logs_dataset
+
+    def gather_files_with_benchmark_tables(self):
+        """Generates file combinations that already have benchmark tables.
+
+        Creates a set of files that already have been loaded to create
+        benchmark tables. Generates list by querying the job.sourceURI field
+        from the results table.
+        Returns:
+            Set of file names that already have been loaded to create
+            benchmark tables.
+        """
+        query = (
+            'SELECT job.sourceURI FROM `{0:s}.{1:s}.{2:s}` '.format(
+                self.bq_project,
+                self.results_table_dataset_id,
+                self.results_table_name,
+            )
+        )
+        query_job = bigquery.Client().query(
+            query,
+            location='US',
+        )
+        files_with_benchmark_tables = set()
+        for row in query_job:
+            if self.bucket_name in row['sourceURI']:
+                uri = row['sourceURI'].split('gs://{0:s}/'.format(
+                    self.bucket_name
+                ))[1]
+                file_name = uri.split('/*')[0]
+                files_with_benchmark_tables.add(file_name)
+        return files_with_benchmark_tables
+
+    def create_benchmark_tables(self):
```

Comment:
I'm trying to figure out how to parallelize some of the other super time consuming processes, like creating tables. I modified the process to: 
```
    def create_benchmark_tables(self):
        """Creates a benchmark table for each file combination in GCS bucket.
        """

        def _create_tables(path_to_load):
            if path_to_load in files_with_benchmark_tables:
                verb = 'Duplicating  '
            else:
                verb = 'Processing '
            logging.info('{0:s} benchmark table for {1:s}'.format(
                verb,
                path_to_load,
            ))
            table = benchmark_table.BenchmarkTable(
                bq_project=self.bq_project,
                gcs_project=self.gcs_project,
                staging_project=self.staging_project,
                staging_dataset_id=self.staging_dataset_id,
                dataset_id=self.dataset_id,
                bucket_name=self.bucket_name,
                path=path_to_load,
                results_table_name=self.results_table_name,
                results_table_dataset_id=self.results_table_dataset_id,
                bq_logs_dataset=self.bq_logs_dataset,
            )
            table.create_table()
            table.load_from_gcs()

        # Gather files combinations that already have benchmark tables.
        files_with_benchmark_tables = self.gather_files_with_benchmark_tables()
        if self.duplicate_benchmark_tables:
            files_to_skip = set()
        else:
            files_to_skip = files_with_benchmark_tables
        # Gather file combinations that exist in the GCS Bucket.
        existing_paths = self.bucket_util.get_existing_paths()
        # Create a benchmark table for each existing file combination, and
        # load the data from the file into the benchmark table.
        paths_to_load = []
        for path in existing_paths:
            path = path.split('/')
            path = '/'.join(path[:len(path) - 1])
            if path not in files_to_skip:
                paths_to_load.append(path)
        with ThreadPoolExecutor() as p:
            p.map(_create_tables, paths_to_load)
```

Which produces the following output: 
```
INFO:root:Done discovering 7 existing files.
INFO:root:Processing  benchmark table for fileType=csv/compression=none/numColumns=10/columnTypes=100_STRING/numFiles=100/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=none/numColumns=10/columnTypes=100_STRING/numFiles=1000/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=none/numColumns=10/columnTypes=100_STRING/numFiles=1/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=gzip/numColumns=10/columnTypes=100_STRING/numFiles=1/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=gzip/numColumns=10/columnTypes=100_STRING/numFiles=1000/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=none/numColumns=10/columnTypes=100_STRING/numFiles=10000/tableSize=2147MB
INFO:root:Processing  benchmark table for fileType=csv/compression=gzip/numColumns=10/columnTypes=100_STRING/numFiles=100/tableSize=2147MB
INFO:root:Created new table: 1564792420
INFO:root:Started load job 29a123d2-6ad0-4351-855e-a7d265d95100 for table 1564792420.

```

So each of the 7 test paths is hitting _create_tables at the same time, but only one load job actually starts running.

Reply:
Nice. I'd recommend using `os.path.split` for your path manipulations here.

Reply:
Is this a question or working as expected?

Reply:
It was an observation, but I should have phrased it as a question haha. It really wasn't working as expected. All 7 files hit the _create_table() function in parallel, but then only one load job would actually proceed. I was throwing it out there in case you had seen anything like it before.

Reply:
Hmm are the load jobs not started or the logging messages not working properly?

Reply:
Unfortunately the load jobs were just starting one at a time. Which was weird. I'm going to dig into this more for the next iteration.

Reply:
the way you're submitting load jobs must be keeping the thread alive making GIL ruin the concurrency.